File: week6/sentimental-credit/credit.py
# Determines whether a provided credit card number is valid according to Luhn’s algorithm

from cs50 import get_string

# Prompts user for the credit card number
while True:
    number = get_string("Number: ")
    if number.isnumeric() and len(number) in [13, 15, 16]:
        break
    else:
        print("INVALID")
        quit()

# Implements Luhn's algorithm to verify mathematical validity of the credit card number
total = 0

for i in range(2, len(number) + 1, 2):
    digit = number[len(number) - i]
    digit = int(digit)
    digit *= 2
    if digit >= 10:
        digit = digit // 10 + digit % 10
    total += digit

for i in range(1, len(number) + 1, 2):
    digit = number[len(number) - i]
    digit = int(digit)
    total += digit

# States whether the card is AMEX, MASTERCARD, VISA, or if the number is invalid
if total % 10 != 0:
    print("INVALID")
elif len(number) == 15 and number[0] == "3" and number[1] in ["4", "7"]:
    print("AMEX")
elif len(number) == 16 and number[0] == "5" and number[1] in ["1", "2", "3", "4", "5"]:
    print("MASTERCARD")
elif len(number) in [13, 16] and number[0] == "4":
    print("VISA")
else:
    print("INVALID")

File: week1/credit/credit.c
// Checks whether the credit card number provided by the user is valid or not, and, if valid, whether it is from AmEx, MasterCard or VISA

#include <cs50.h>
#include <stdio.h>

long get_number(void);
string algorithm(long number);

int main(void)
{
    long number = get_number();
    string result = algorithm(number);
    printf("%s", result);
}

// Prompts user for input
long get_number(void)
{
    long number;
    do
    {
        number = get_long("Please insert a credit card number (no hyphens): ");
    }
    while (number < 0);
    return number;
}

// Checks if the number provided can be validated mathematically, and to which company it is associated
string algorithm(long number)
{
    int accumulator = 0;
    long position;
    long digit;

    // Starting from the decimal position, aka the second-to-last digit in the credit card number
    position = 10;
    while (number / position >= 1)
    {
        // Gets the digit associated with the current position
        digit = (number % (position * 10)) / position;
        digit *= 2;
        // Transforms a 2-digit number into the sum of its 2 digits
        if (digit >= 10)
        {
            digit = (digit / 10) + (digit % 10);
        }
        accumulator += digit;
        // Goes 2 digits to the left
        position *= 100;
    }

    // Starting again from the unity position, aka the last digit in the credit card number
    position = 1;
    while (number / position >= 1)
    {
        // Gets the digit associated with the current position
        digit = (number % (position * 10)) / position;
        accumulator += digit;
        // Goes 2 digits to the left
        position *= 100;
    }

    if (((accumulator % 10) != 0) || (number > 5599999999999999) || (number < 4000000000000))
    {
        return "INVALID\n";
    }
    else
    {
        // In case it is a mathematically valid credit card number
        if ((number / (position / 100) == 4) || (number / (position / 10) == 4))
        {
            return "VISA\n";
        }
        else if ((number / (position / 1000) == 34) || (number / (position / 100) == 34) ||
                 (number / (position / 1000) == 37) || (number / (position / 100) == 37))
        {
            return "AMEX\n";
        }
        else if ((number / (position / 1000) == 51) || (number / (position / 100) == 51) ||
                 (number / (position / 1000) == 52) || (number / (position / 100) == 52) ||
                 (number / (position / 1000) == 53) || (number / (position / 100) == 53) ||
                 (number / (position / 1000) == 54) || (number / (position / 100) == 54) ||
                 (number / (position / 1000) == 55) || (number / (position / 100) == 55))
        {
            return "MASTERCARD\n";
        }
        else
        {
            return "INVALID\n";
        }
    }
}